substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/PreMainSupport.java

```diff
@@ -0,0 +1,234 @@
+ * Copyright (c) 2024, 2024, Alibaba Group Holding Limited. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.core;
+
+import com.oracle.svm.core.util.VMError;
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+
+import java.lang.instrument.ClassDefinition;
+import java.lang.instrument.ClassFileTransformer;
+import java.lang.instrument.Instrumentation;
+import java.lang.instrument.UnmodifiableClassException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.jar.JarFile;
+
+/**
+ * Java agent can do instrumentation initialization work in premain phase. This class supports
+ * registering such premain methods at native image build time and invoking them at native image
+ * runtime. <br>
+ * JVM supports two kind of premain methods:
+ * <ol>
+ * <li>{@code public static void premain(String agentArgs, Instrumentation inst)}</li>
+ * <li>{@code public static void premain(String agentArgs)}</li>
+ * </ol>
+ * For the first one, at registration time we will set the second parameter with an instance of
+ * {@link SVMRuntimeInstrumentImpl} class which does not do any actual work as no instrumentation
+ * can do in native code at runtime.
+ * <p>
+ * <b>Be noticed</b>, the original agent premain method may not work well at native image runtime
+ * even if the input {@link Instrumentation} class is replaced with
+ * {@link SVMRuntimeInstrumentImpl}.
+ * </p>
+ * <p>
+ * It is the agent developers' responsibility to implement a native version of their agent premain
+ * method. It can be implemented in two ways:
+ * <ul>
+ * <li>Isolate code by checking current runtime. For example: <code>
+ *         <pre>
+ *         String vm = System.getProperty("java.vm.name");
```

Comment:
We should guide the users to use the properties from [ImageInfo](https://github.com/oracle/graal/blob/2449afbb28dadf2d9a70a866da06d664f1d1a539/sdk/src/org.graalvm.nativeimage/src/org/graalvm/nativeimage/ImageInfo.java#L67). Then they can see if the agent is applied at build time or at runtime. We also need to make sure that this property is set before premain is called for the Native Image builder.

Reply:
Fixed. This property can be obtained in `premain`.